src/utils.py

```diff
@@ -386,6 +386,107 @@ def powerset(seq: Sequence, exclude_empty: bool) -> Iterator[Sequence]:
 
+_S = TypeVar('_S', bound=Hashable)  # state in heuristic search
+_A = TypeVar('_A')  # action in heuristic search
+
+
+@dataclass(frozen=True)
+class _HeuristicSearchNode(Generic[_S, _A]):
+    state: _S
+    edge_cost: float
+    cumulative_cost: float
+    parent: Optional[_HeuristicSearchNode[_S, _A]] = None
+    action: Optional[_A] = None
+
+
+def _run_heuristic_search(
+    initial_state: _S,
+    check_goal: Callable[[_S], bool],
+    get_successors: Callable[[_S], Iterator[Tuple[_A, _S, float]]],
+    get_priority: Callable[[_HeuristicSearchNode[_S, _A]], Any],
+    max_expansions: int = 1000
+    ) -> Tuple[List[_S], List[_A]]:
+    """A generic heuristic search implementation.
+
+    Depending on get_priority, can implement A*, GBFS, or UCS.
+
+    If no goal is found, returns the node with the best priority.
+    """
+    queue: List[Tuple[Any, int, _HeuristicSearchNode[_S, _A]]] = []
+    frontier = set()
+    explored = set()
+
+    root_node: _HeuristicSearchNode[_S, _A] = _HeuristicSearchNode(
+        initial_state, 0, 0)
+    root_priority = get_priority(root_node)
+    best_node = root_node
+    best_node_priority = root_priority
+    tiebreak = itertools.count()
+    hq.heappush(queue, (root_priority, next(tiebreak), root_node))
+    frontier.add(initial_state)
+    num_expansions = 0
+
+    while len(queue) > 0 and num_expansions < max_expansions:
+        _, _, node = hq.heappop(queue)
+        # If the goal holds, return.
+        if check_goal(node.state):
+            return _finish_plan(node)
+        # Add node state to explored set.
+        explored.add(node.state)
+        num_expansions += 1
+        # Generate successors.
+        for action, child_state, cost in get_successors(node.state):
+            # If the state is already in explored or frontier, don't bother.
+            if child_state in frontier | explored:
+                continue
+            # Add new node
+            child_node = _HeuristicSearchNode(
+                state=child_state,
+                edge_cost=cost,
+                cumulative_cost=node.cumulative_cost+cost,
+                parent=node,
+                action=action)
+            priority = get_priority(child_node)
+            hq.heappush(queue, (priority, next(tiebreak), child_node))
+            frontier.add(child_state)
+            if priority < best_node_priority:
+                best_node_priority = priority
+                best_node = child_node
+
+    # Did not find path to goal; return best path seen.
+    return _finish_plan(best_node)
+
+
+def _finish_plan(node: _HeuristicSearchNode[_S, _A]
+                 ) -> Tuple[List[_S], List[_A]]:
+    """Helper for _run_heuristic_search.
+    """
+    rev_state_sequence: List[_S] = []
+    rev_action_sequence: List[_A] = []
+
+    while node.parent is not None:
+        action = cast(_A, node.action)
+        rev_action_sequence.append(action)
+        rev_state_sequence.append(node.state)
+        node = node.parent
```

Comment:
shouldn't the returned state_sequence be of length 1 more than the returned action_sequence?

Reply:
oops that was a dumb one